tools/please_pex/preamble/path.c

```diff
@@ -0,0 +1,234 @@
+ * Otherwise, get_plz_bin_path returns the path to the bin/ subdirectory within the plz-out/ directory
+ * to which this .pex file belongs. If the .pex file does not exist within a plz-out/ directory
+ * tree, get_plz_bin_path fails.
+ */
+err_t *get_plz_bin_path(char **path) {
+    char *pex_dir = NULL;
+    char *tmp_dir = NULL;
+    char *pex_realpath = NULL;
+    char *tmp_realpath = NULL;
+    size_t pex_len = 0;
+    size_t tmp_len = 0;
+    err_t *err = NULL;
+
+    if ((err = get_pex_dir(&pex_dir)) != NULL) {
+        err = err_wrap("get_pex_dir", err);
+        goto end;
+    }
+
+    if (getenv("PLZ_ENV") != NULL) {
+        if ((tmp_dir = getenv("TMP_DIR")) == NULL) {
+            // There are two circumstances under which this might happen: either if this is a Please
+            // build environment but TMP_DIR was unset before the .pex file was executed, or if this
+            // isn't a Please build environment and PLZ_ENV just happened to be set for some other
+            // reason. The latter is more likely than the former (many build definitions rely on
+            // TMP_DIR being defined, which would lead to much more breakage besides this function
+            // failing) - therefore, optimise for the latter case.
+            log_warn("PLZ_ENV is defined but TMP_DIR is not; assuming this is not a Please build environment");
+            goto no_plz_env;
+        }
+
+        // Identify whether the .pex file exists inside the build environment.
+        if ((pex_realpath = realpath(pex_dir, NULL)) == NULL) {
+            err = err_from_errno("realpath pex_dir");
+            goto end;
+        }
+        if ((tmp_realpath = realpath(tmp_dir, NULL)) == NULL) {
+            err = err_from_errno("realpath tmp_dir");
+            goto end;
+        }
+
+        pex_len = strlen(pex_realpath);
+        tmp_len = strlen(tmp_realpath);
+
+        if (
+            strncmp(tmp_realpath, pex_realpath, tmp_len) == 0 &&
+            (
+                (pex_len == tmp_len) ||
+                (pex_len - tmp_len >= 1 && pex_realpath[tmp_len] == '/')
+            )
+        ) {
+            if (((*path) = strdup(tmp_realpath)) == NULL) {
+                err = err_from_errno("strdup");
+            }
+            goto end;
+        }
+    }
+
+no_plz_env:
+    // Try searching the .pex file path component-wise for the last instance of "plz-out". If one is
+    // found, assume the binary outputs are in the bin/ subdirectory within that directory. There's
+    // no need to check now whether bin/ actually exists - if it doesn't, the call to execvp() later
+    // will fail anyway.
+    while (!STREQ(basename(pex_dir), "plz-out") && !STREQ(pex_dir, "/") && !STREQ(pex_dir, ".")) {
+        pex_dir = dirname(pex_dir);
+    }
+    if (STREQ(pex_dir, "/") || STREQ(pex_dir, ".")) {
+        // If we reached the left-most component in the path, the .pex file doesn't exist within a
+        // plz-out/ directory tree.
+        goto end;
```

Comment:
Do we need to make an error here? The docstring says that get_plz_bin_path fails if we can't determine a path.

Reply:
Yes - this is a holdover from before I wrote `pexerror.c`, when I was doing error handling in a much more complicated way. Thanks, I'll set an error here.